File: frontend/src/types/document.ts
/**
 * Document types matching backend models
 */

export type DocumentType = 'invoice' | 'insurance_claim' | 'compliance_form' | 'unknown';

export type DocumentStatus =
  | 'uploaded'
  | 'preprocessing'
  | 'ocr_processing'
  | 'extracting'
  | 'validating'
  | 'needs_review'
  | 'in_review'
  | 'completed'
  | 'failed';

export interface DocumentMetadata {
  source?: string;
  sender?: string;
  [key: string]: unknown;
}

export interface Document {
  id: string;
  filename: string;
  document_type: DocumentType;
  status: DocumentStatus;
  storage_path: string;
  page_count: number;
  file_size: number;
  metadata: DocumentMetadata;
  created_at: string;
  updated_at: string;
}

export interface DocumentUploadRequest {
  file: File;
  document_type?: DocumentType;
  metadata?: DocumentMetadata;
}

export interface DocumentUploadResponse {
  id: string;
  filename: string;
  document_type: DocumentType;
  status: DocumentStatus;
  created_at: string;
  estimated_completion?: string;
}

export interface DocumentListParams {
  page?: number;
  page_size?: number;
  document_type?: DocumentType;
  status?: DocumentStatus;
  search?: string;
  sort_by?: 'created_at' | 'filename' | 'status';
  sort_order?: 'asc' | 'desc';
}

export interface DocumentListResponse {
  items: Document[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface DocumentStatusResponse {
  id: string;
  status: DocumentStatus;
  progress?: number;
  current_step?: string;
  error_message?: string;
}

// Helper functions
export function isProcessing(status: DocumentStatus): boolean {
  return ['uploaded', 'preprocessing', 'ocr_processing', 'extracting', 'validating'].includes(status);
}

export function isTerminal(status: DocumentStatus): boolean {
  return ['completed', 'failed'].includes(status);
}

export function needsReview(status: DocumentStatus): boolean {
  return ['needs_review', 'in_review'].includes(status);
}

export function getStatusLabel(status: DocumentStatus): string {
  const labels: Record<DocumentStatus, string> = {
    uploaded: 'Uploaded',
    preprocessing: 'Preprocessing',
    ocr_processing: 'OCR Processing',
    extracting: 'Extracting',
    validating: 'Validating',
    needs_review: 'Needs Review',
    in_review: 'In Review',
    completed: 'Completed',
    failed: 'Failed',
  };
  return labels[status];
}

export function getStatusColor(status: DocumentStatus): string {
  const colors: Record<DocumentStatus, string> = {
    uploaded: 'status-uploaded',
    preprocessing: 'status-processing',
    ocr_processing: 'status-processing',
    extracting: 'status-processing',
    validating: 'status-processing',
    needs_review: 'status-needs-review',
    in_review: 'status-needs-review',
    completed: 'status-completed',
    failed: 'status-failed',
  };
  return colors[status];
}

export function getDocumentTypeLabel(type: DocumentType): string {
  const labels: Record<DocumentType, string> = {
    invoice: 'Invoice',
    insurance_claim: 'Insurance Claim',
    compliance_form: 'Compliance Form',
    unknown: 'Unknown',
  };
  return labels[type];
}

File: backend/src/docpipeliner/models/document.py
"""Document-related data models."""

from datetime import datetime
from enum import Enum
from typing import Any, Optional
from uuid import UUID, uuid4

from pydantic import BaseModel, Field


class DocumentType(str, Enum):
    """Supported document types for extraction."""

    INVOICE = "invoice"
    INSURANCE_CLAIM = "insurance_claim"
    COMPLIANCE_FORM = "compliance_form"
    UNKNOWN = "unknown"


class DocumentStatus(str, Enum):
    """Document processing status."""

    UPLOADED = "uploaded"
    PREPROCESSING = "preprocessing"
    OCR_PROCESSING = "ocr_processing"
    EXTRACTING = "extracting"
    VALIDATING = "validating"
    NEEDS_REVIEW = "needs_review"
    IN_REVIEW = "in_review"
    COMPLETED = "completed"
    FAILED = "failed"


class BoundingBox(BaseModel):
    """Bounding box coordinates for extracted text."""

    x: float = Field(..., description="X coordinate (left)")
    y: float = Field(..., description="Y coordinate (top)")
    width: float = Field(..., ge=0, description="Width of bounding box")
    height: float = Field(..., ge=0, description="Height of bounding box")
    page: int = Field(..., ge=1, description="Page number (1-indexed)")


class PageCreate(BaseModel):
    """Schema for creating a new page record."""

    document_id: UUID
    page_number: int = Field(..., ge=1)
    image_path: str
    classification: Optional[str] = None


class Page(PageCreate):
    """Page model with database fields."""

    id: UUID = Field(default_factory=uuid4)
    created_at: datetime = Field(default_factory=datetime.utcnow)

    class Config:
        from_attributes = True


class DocumentCreate(BaseModel):
    """Schema for creating a new document."""

    filename: str = Field(..., min_length=1, max_length=255)
    document_type: DocumentType = DocumentType.UNKNOWN
    metadata: dict[str, Any] = Field(default_factory=dict)


class Document(BaseModel):
    """Document model with all fields."""

    id: UUID = Field(default_factory=uuid4)
    filename: str
    document_type: DocumentType
    storage_path: str
    status: DocumentStatus = DocumentStatus.UPLOADED
    metadata: dict[str, Any] = Field(default_factory=dict)
    page_count: Optional[int] = None
    file_size_bytes: Optional[int] = None
    error_message: Optional[str] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

    class Config:
        from_attributes = True


class DocumentResponse(BaseModel):
    """API response schema for document."""

    id: UUID
    filename: str
    document_type: DocumentType
    status: DocumentStatus
    page_count: Optional[int] = None
    file_size_bytes: Optional[int] = None
    metadata: dict[str, Any] = Field(default_factory=dict)
    error_message: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    estimated_completion: Optional[datetime] = None

    class Config:
        from_attributes = True
